### PR TITLE
Add threadsafe method to get featuresource from layer

### PR DESCRIPTION
**Pull request by m-kuhn**
Adds a thread-safe method to get a feature source given a weak map layer pointer.
Works only starting in a safe manner with Qt >= 5.10.
Opinions on that?

### CODE REVIEW

**Comment by elpaso** [comment]
looks good, just curious what's new in QT 5.10 within this regard?

**Comment by m-kuhn** [comment]
This one:
http://doc.qt.io/qt-5/qmetaobject.html#invokeMethod-5

File: src/core/qgsvectorlayerutils.cpp [hunk 1]
```diff
@@ -498,6 +498,34 @@ QgsFeature QgsVectorLayerUtils::duplicateFeature( QgsVectorLayer *layer, const Q
   return newFeature;
 }
 
+std::unique_ptr<QgsVectorLayerFeatureSource> QgsVectorLayerUtils::getFeatureSource( QWeakPointer<QgsVectorLayer> layer )
+{
+  std::unique_ptr<QgsVectorLayerFeatureSource> featureSource;
+
+  auto getFeatureSource = [ layer, &featureSource ]
+  {
+    QgsVectorLayer *lyr = layer.data();
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerutils.cpp]: Could you add an assert in here (for qt 10 and above) checking that it's the main thread?

And then with a simple unit test checking that a source is returned we should be protected here.

File: src/core/qgsvectorlayerutils.h [hunk 2]
```diff
@@ -155,6 +156,17 @@ class CORE_EXPORT QgsVectorLayerUtils
      */
     static QgsFeature duplicateFeature( QgsVectorLayer *layer, const QgsFeature &feature, QgsProject *project, int depth, QgsDuplicateFeatureContext &duplicateFeatureContext SIP_OUT );
 
+    /**
+     * Gets the feature source from a weak QgsVectorLayer pointer.
+     * This method is thread-safe but will block the main thread for execution. Executing it from the main
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerutils.h]: This isn't quite true, right? It makes it sound like the main thread is blocked for the duration of the call -- but it's only blocked for the duration of getting the source, right? The rest of the the the main thread is humming away doing whatever else it's up to.

Comment (m-kuhn) [reply src/core/qgsvectorlayerutils.h]: if by "the call" you mean "consuming the iterator", then your statement is true. Are you proposing to replace the word "execution" with something else?

Comment (nyalldawson) [reply src/core/qgsvectorlayerutils.h]: Yeah, but honestly I can't come up with an elegant way to word this!

**Comment by nyalldawson** [comment]
I've been trying to think of there's a way we could do this for earlier qt (which is the biggest limitation with rolling out this method everywhere it's currently needed)

Looks like there's some options in https://stackoverflow.com/questions/21646467/how-to-execute-a-functor-or-a-lambda-in-a-given-thread-in-qt-gcd-style

**Comment by m-kuhn** [comment]
I'll leave this task to people that stick to systems with older Qt versions.

**Comment by nyalldawson** [comment]
Fine with me. If you add the assert and a test then merge and I'll look at the compatibility code

**Comment by m-kuhn** [comment]
Ready to merge

**Review by nyalldawson — APPROVED** [review]
LGTM

**Pull request merged**